File: LeetCode/재귀 3 문자열 디코딩.py
# 재귀 이용한 풀이법
# 스택을 사용하면 더 쉽고 빠름
class Solution:
    def decodeString(self, s: str) -> str:
        # 대괄호 매칭 딕셔너리
        # 각 '['에 맞는 ']'의 위치를 계산
        close = {}
        st = []
        for i, c in enumerate(s):
            if c == '[':
                st.append(i)
            elif c == ']':
                close[st.pop()] = i
        # dfs(l, r) 함수
        def solve(l, r):
            # 반복 횟수 저장
            num = 0
            # 문자열 단어 조각 저장
            ans = []
            while l <= r:
                c = s[l]
                # s[i]가 숫자면 num에 계속 누적하기
                if c.isdigit():
                    num = num*10 + int(c)
                # s[i]가 앞에서 읽은 숫자 num만큼 괄호 안 문자열을 반복
                # 문자열을 재귀적으로 디코딩
                # 카운트를 새로 시작
                # 닫는 위치로 인덱스 점프
                elif c == '[':
                    ans.append(num*solve(l+1, close[l]-1))
                    num = 0
                    l = close[l]
                # s[i]가 문자면 현재 임시 단어에 더하기
                else:
                    ans.append(c)
                l += 1
            # 생성된 모든 단어들을 하나로 합치기
            return "".join(ans)
        return solve(0, len(s)-1)

File: LeetCode/재귀 3 문자열 디코딩.cpp
class Solution {
public:
    string decodeString(string s) {
        // 시작 위치
        int pos = 0;
        return helper(pos, s);
    }

    string helper(int& pos, string s) {
        // 각 단어 블록의 반복해야 하는 횟수
        int num = 0;
        // 전체 문자열
        string word = "";
        for (; pos < s.size(); pos++) {
            char cur = s[pos];
            // 현재 기호가 '['인 경우
            // 새로운 단어 구성을 탐색 시작해서 단어를 새로 완성
            // 해당 단어를 num만큼 전체 문장에 합치기
            if (cur == '[') {
                string cur_str = helper(++pos, s);
                for (; num > 0; num--) {
                    word += cur_str;
                }
            // 현재 기호가 숫자인 경우
            // 해당 블록의 전체 수를 구성
            } else if (cur >= '0' && cur <= '9') {
                num = num*10 + (cur - '0');
            // 현재 기호가 ']'인 경우
            // 새로운 단어를 완성하고 반환
            } else if (cur == ']') {
                return word;
            // 현재 기호가 문자인 경우
            // 새로운 단어를 계속 구성해나감
            } else {
                word += cur;
            }
        }
        return word;
    }
};